### DISCUSSION TITLE
Future handling of analog switches

### DISCUSSION THREAD

**Discussion by rotorman** [EdgeTX/edgetx#1450] [Ideas]
Presently, e.g. 6 position on a number of radios (TX16S, T16, T18), some 2 and 3 position switches (e.g. NV14, PL18, PL18EV) or button arrays (e.g PL18EV stick end buttons), are being included in CalibratedAnalogs and handled as proportional analog channels.

E.g. see:
https://github.com/EdgeTX/edgetx/blob/8150b00823ca0a37ade72c41bb2fa35cbe31e594/radio/src/targets/horus/board.h#L448 ,  https://github.com/EdgeTX/edgetx/blob/8150b00823ca0a37ade72c41bb2fa35cbe31e594/radio/src/targets/nv14/board.h#L324-L331 and https://github.com/EdgeTX/edgetx/blob/a704b575e08b297b34e887b192a18aa2f3939763/radio/src/targets/pl18/board.h#L335-L340

All these analog switch options have predetermined output values, with some manufacturing tolerance, and in my opinion can be skipped from calibration. Whereas on 2 and 3 position switches, the output is either near 0V, near VREF, or near ratiometric middle (3-pos), the division is not so equidistant for the 6-pos. on TX16S, T16, T18 and the 4-pos. PL18EV stick end buttons (4 positions, as the analog output on PL18EV stick end buttons has 4 levels: no button pressed, 1st button pressed, 2nd button pressed, both buttons pressed simultaneously). Nevertheless, also for these two latter special cases (6 and 4 pos.), the values provided from the radios are known, and with the help of a community, the manufacturing deviations can be estimated.

The suggestion is to drop the calibration requirement for analog switches and going with pre-programmed thresholds for the detection (for more fancy cases, like the 6-pos and the 4-pos, the thresholds likely need to be hard-coded to be specific to the radio target).
Here, a community feedback (e.g. with the help of PR #1421) could be useful to figure the ideal borders.

Consider following fictive 4-position analog switch that from a range of tested radios would give a following normalized distribution for the 4 switches across a 12-bit ADC range:

![grafik](https://user-images.githubusercontent.com/21011587/149657791-aec3b27a-7a18-44c9-95b9-150f3d966877.png)

The means of switch positions in 12-bit raw values are:
switch 1 = 50
switch 2 = 1500
switch 3 = 2200
switch 4 = 3400
In this fictive Gaussian distribution example, no overlap happens and perfect thresholding is possible.
The detection borders could in this fictive case be:
A = 775 (= (1500 - 50)/2 + 50 )
B = 1850
C = 2800 

![grafik](https://user-images.githubusercontent.com/21011587/149657825-351eb728-a4f2-4c92-97aa-77f3e2af2456.png)

Output in blue zone would be detected as switch 1, orange as switch 2, gray as sw3 and yellow as sw 4.

In order to facilitate outliers and possible own hardware modifications that might deviate from the standard separation values, the idea would be to allow optional calibration values in YAML to be used, e.g. similar to `txVoltageCalibration` to override or offset the hard-coded default values.

Interested on your thoughts! Thx!

**Comment by eshifri** [comment]
1. Just for my education: I have never seen calibration option for switches on NV14. Where is it?
2. If there is a way to put calibration into YAML, what is the point of hard-coding them? Why not write "hardcoded" defaults into YAML?
3. I think it was mentioned on Discord: analog outputs should be available on the HW diagnostic screen.

**Reply by pfeerick** [reply]
Calibration is for analogs, not for switches. What Risto is referring to is the hybrid 'analog switches', i.e. the 6POS on the TX16, which is a single analog input that reports one of six positions in the range depending on which button you press. 

I do have to wonder why that exists in the NV14 config - I guess there most be some other part of the code that matches it to being a switch thus uses a combination of the switch configuration (toggle, 2POS, 3POS) + predetermined range to handle them without actual calibration needed?

I also think the term hard-coded is a bit misleading - I also think it should simply be default values written into the yaml - thus it can be tweaked in the off-chance it needs it (if a UI to edit those values isn't added at some point. Or more specifically, when you choose 6POS, those should be the default values that are written.

**Comment by eshifri** [comment]
Sorry I was not clear enough: my question was about NV14 switches; I understand the reasons behind the 6POS calibration.
I suspect the switch calibration for NV14 is not in use.

**Comment by rotorman** [comment]
On NV14, all switches are connected to analog inputs. I assume, this design decision was made to be able to have more than 2-pos switch with only a single pin on the STM32F4 microcontroller.

The code presently uses rigid thresholds for detecting the switch positions on NV14:
https://github.com/EdgeTX/edgetx/blob/dcbdde2b74baa3398306d06021b7bf7a65885efa/radio/src/targets/nv14/keys_driver.cpp#L128-L141

The raw 12-bit ADC value range goes from 0 to 4095. A measurement with a value less than 1024 is output as position 0, measurements between 1024 and 3072 are output as position 1 and measurements above 3072 as position 2.

The analog switches, although not included in, nor strictly necessary for calibration, are wrongly listed under `CalibratedAnalogs` and as thus `NUM_CALIBRATED_ANALOGS` is not really correct:
https://github.com/EdgeTX/edgetx/blob/8150b00823ca0a37ade72c41bb2fa35cbe31e594/radio/src/targets/nv14/board.h#L317-L333

The `NUM_CALIBRATED_ANALOGS` is to define the depth of a global variable that is further used in various parts of the code:
https://github.com/EdgeTX/edgetx/blob/dcbdde2b74baa3398306d06021b7bf7a65885efa/radio/src/mixer.cpp#L38
